modules/javafx.graphics/src/main/java/com/sun/javafx/css/FixedCapacitySet.java

```diff
@@ -0,0 +1,582 @@
+ * and only suitable for very specific use cases where the maximum size is known in advance
+ * and when items are expected to never be removed from the sets. Fixed capacity sets will
+ * throw {@link IllegalStateException} if adding an element would exceed the maximum capacity.
+ *
+ * <p>These sets do not allow {@code null} elements, and unless otherwise specified, passing
+ * {@code null} for any argument will result in a {@link NullPointerException}.
+ *
+ * <p>Specifically, these sets are optimized for holding sets of style class names, which
+ * have the following characteristics:
+ *
+ * <ul>
+ * <li>Never {@code null}</li>
+ * <li>Sets in almost all cases contain just 1 or 2 elements, only rarely containing 3 or more</li>
+ * <li>Sets are often compared using containsAll; to avoid creating an iterator when the input is not
+ *   a {@code FixedSizeSet}, the inverse function {@link #isSuperSetOf(Collection)} is provided</li>
+ * </ul>
+ *
+ * The provided implementations are optimized for memory use, fast containsAll and
+ * fast iteration. Generally, these sets will use half the memory of an equivalent
+ * {@code HashSet} (and comparable to the immutable sets provided by {@code Set.of()})
+ * as they do not require a wrapper to hold each element.
+ *
+ * <p>These sets can only be appended, reject {@code null}s, have a fixed maximum size (which will throw
+ * an exception if exceeded), and can be frozen (made read-only without using a wrapper).
+ *
+ * <p>The fall back set implementation for large sets uses open addressing. It is
+ * only lightly optimized as the expectation is that it will see little to no use.
+ * It is still preferred over {@code HashSet} due to its low memory foot print, and
+ * faster iteration.
+ *
+ * @param <T> the element type
+ */
+public sealed abstract class FixedCapacitySet<T> extends AbstractSet<T> {
+    private static final FixedCapacitySet<?> EMPTY;
+
+    static {
+        EMPTY = new Single<>();
+
+        EMPTY.freeze();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> FixedCapacitySet<T> empty() {
+        return (FixedCapacitySet<T>) EMPTY;
+    }
+
+    /**
+     * Creates a new {@link FixedCapacitySet} with the given maximum capacity.
+     * If the capacity is exceeded, fixed capacity sets do not grow, but instead
+     * throw an {@link IllegalStateException}.
+     *
+     * @param <T> the element type
+     * @param maximumCapacity the maximum possible number of elements the set can hold, cannot be negative
+     * @return a new empty set, never {@code null}
+     */
+    public static <T> FixedCapacitySet<T> of(int maximumCapacity) {
+        return maximumCapacity == 0 ? empty()
+             : maximumCapacity == 1 ? new Single<>()
+             : maximumCapacity == 2 ? new Duo<>()
+             : maximumCapacity < 10 ? new Hashless<>(maximumCapacity)  // will reject negative values
+                                    : new OpenAddressed<>(maximumCapacity);
```

Comment:
It's possible to use the standard `HashSet` as a final fallback (or even to replace `OpenAddressed`), although it would need to be wrapped so support the `freeze` method (or we need to implement this differently).

There are a few reasons however not to do so:
- `HashSet` doesn't support `freeze` so we'd need to copy it after it is created; we can't use `Set.of` as this would also require making a copy of the entries first (the entries are added one by one because empty values are filtered) -- the solution in this PR carefully avoids all unnecessary copying and allocations to achieve the better performance
- The sets are indeed in almost all cases very small (hence the optimizations for small sets) -- they contain style classes that apply to a single node or a single selector.  Nodes with many style classes applied to them (ie. 10 or more) I think rarely occurs in practice. Same goes for selectors; you'd have to have something like `a > b c d e f g h i > j` as selector -- remember that this doesn't apply to something like `a, b, c, d, e, f, g, h, i, j`, as they are actually 10 different selectors with 1 style each.
- `HashSet` is less memory efficient (between 21.5 to 26.7 bytes per entry with the load factor varying between 0.75 and 0.75 * 0.5 and a base cost of 16 bytes per entry) vs `OpenAddressed` which requires between 5.7 and 13.3 bytes per entry (due to its load factor variation between 0.3 and 0.7 with a base cost of 0 bytes per entry) -- why isn't the JDK using open addressed then you may wonder? Well, it actually does in its `Set.of` implementation, but not for `HashSet`, primarily because `HashSet` has less worse edge cases when hash codes are poor, and it has to be more generally usable (open addressed tables degrade easier and require more maintenance when under constant modification, which doesn't apply for us here).

Reply:
makes sense, thanks!